rfcs/text/0009_screenshot_mode_service.md

```diff
@@ -0,0 +1,262 @@
+performance for Reporting if we only look for ways to change the Reporting code
+to do it. In other words, it will take a cross-team effort to improve the main
+performance issues. This RFC proposes one way to make cross-team efforts work
+smoothly, which is to have a new plugin-provided service in Kibana that helps UIs and
+visualizations work nicely when loaded for screenshot capture.
+
+**Screenshot mode service**
+
+Applications in Kibana should be aware of when their rendering is for a human
+user using the page interactively, and when it's not. Sometimes the page is
+loaded for display-purposes: there is no user in control of the
+browser and no need for interactivity. Having Kibana support that purpose
+allows higher-level features to be built on top of rapidly-available images of
+Kibana pages. This turns out to be crucial for features like scheduled PDF
+report generation, and automated alerts with embedded images.
+
+As a bridge between the Kibana applications screenshot-capture features is
+needed as a service in Kibana. This RFC names that service the Screenshot
+Mode Service (name TBD). Applications would read from this service while
+loading UI components into the browser, and help the application make
+case-by-case bases for loading components and the parameters that get passed
+into the components.
+
+In most cases, the information coming from this service would help the
+application determine which UI components should **not** be loaded, as applications will
+use the service response for skipping components that aren't needed for
+presenting the data in a screenshot, and thus work towards better performance
+for the display-only purpose.
+
+More background on how Reporting currently works, including the lifecycle of
+creating a PNG report, is here: https://github.com/elastic/kibana/issues/59396
+
+# Basic example
+
+When Kibana loads initially, there is a Newsfeed component in the UI that
+checks internally cached records to see if it must fetch the Elastic News
+Service for newer items. When the Screenshot Mode Service is implemented, the
+Newsfeed component has a source of information to check on whether or not it
+should load in the Kibana UI. If it can avoid loading, it avoids an unnecessary
+HTTP round trip, which weigh heavily on performance.
+
+# Motivation
+
+Kibana Reporting is a commercial feature in Elastic and is highly capable of
+loading the application pages and converting them into a screenshot for PNG or
+PDF export. However, the way it works has downsides with performance,
+maintainability, and expanding it as a tool that can power higher-level
+features. The solution to those downsides is to have Kibana pages themselves
+become more capable at presenting themselves for screenshot capture reports.
+With the Screenshot Mode Service available, Reporting could drop the
+task that it currently has which hurt performance: wasted rendering that is
```

Comment:
According to [your investigation](https://github.com/elastic/kibana/issues/59396#issuecomment-598415734). 
there are 2 steps being performance bottlenecks in the PDF rendering:
- `open_url`. which is hurt by bloated artifacts size.
- `get_number_of_items`. I don't think we spend several seconds accessing DOM. Do we wait for data to be fetched on this step?

How would the current proposal improve them?

Reply:
* To improve the performance of open_url, we should add handling for screenshot mode on the server side, in the route to eliminate unnecessary code or plugins on loading
* To improve the performance of `get_number_of_items` (and `wait_for_render`), we'll look for ways to streamline the UX of the applications for screenshot capture mode to get them to be in a "ready" state sooner.

Reply:
> I don't think we spend several seconds accessing DOM. Do we wait for data to be fetched on this step?

We're waiting for the application to inject the meta data into the DOM, so the time we're spending here is waiting for the DOM selectors to yield a result.